File: android/demo/src/main/kotlin/app/mekuri/demo/DemoFigure.kt
package app.mekuri.demo

import androidx.compose.foundation.Canvas
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.scale

enum class DemoPose {
    Standing,
    Pointing,
    Running,
    Reading,
}

/**
 * A silhouette drawn from a head, a torso and stroked limbs, scaled to the
 * height it is given. [mirrored] flips it about its own centre.
 */
@Composable
fun DemoFigure(pose: DemoPose, modifier: Modifier = Modifier, mirrored: Boolean = false) {
    Canvas(modifier) {
        if (mirrored) {
            scale(scaleX = -1f, scaleY = 1f) { drawFigure(pose) }
        } else {
            drawFigure(pose)
        }
    }
}

private fun DrawScope.drawFigure(pose: DemoPose) {
    val unit = this.size.height
    fun at(x: Float, y: Float) = Offset(this.size.width / 2 + x * unit, y * unit)

    val lean = if (pose == DemoPose.Running) 0.06f else 0f
    val headRadius = unit * 0.10f
    drawCircle(DemoInk.Ink, radius = headRadius, center = at(lean, 0.13f))

    val torso = Path().apply {
        val a = at(-0.10f + lean, 0.26f)
        val b = at(0.10f + lean, 0.26f)
        val c = at(0.08f, 0.58f)
        val d = at(-0.08f, 0.58f)
        moveTo(a.x, a.y)
        lineTo(b.x, b.y)
        lineTo(c.x, c.y)
        lineTo(d.x, d.y)
        close()
    }
    drawPath(torso, DemoInk.Ink)

    val limbs = Path()
    fun segment(x1: Float, y1: Float, x2: Float, y2: Float) {
        val from = at(x1, y1)
        val to = at(x2, y2)
        limbs.moveTo(from.x, from.y)
        limbs.lineTo(to.x, to.y)
    }
    when (pose) {
        DemoPose.Standing -> {
            segment(-0.09f, 0.30f, -0.16f, 0.55f)
            segment(0.09f, 0.30f, 0.16f, 0.55f)
            segment(-0.06f, 0.58f, -0.08f, 0.95f)
            segment(0.06f, 0.58f, 0.08f, 0.95f)
        }
        DemoPose.Pointing -> {
            segment(-0.09f, 0.30f, -0.15f, 0.55f)
            segment(0.09f, 0.30f, 0.36f, 0.30f)
            segment(-0.06f, 0.58f, -0.08f, 0.95f)
            segment(0.06f, 0.58f, 0.10f, 0.95f)
        }
        DemoPose.Running -> {
            segment(-0.04f, 0.30f, -0.24f, 0.42f)
            segment(0.14f, 0.30f, 0.30f, 0.18f)
            segment(-0.06f, 0.58f, -0.30f, 0.82f)
            segment(0.06f, 0.58f, 0.26f, 0.92f)
        }
        DemoPose.Reading -> {
            segment(-0.09f, 0.30f, -0.12f, 0.46f)
            segment(0.09f, 0.30f, 0.12f, 0.46f)
            segment(-0.06f, 0.58f, -0.08f, 0.95f)
            segment(0.06f, 0.58f, 0.08f, 0.95f)
        }
    }
    drawPath(
        limbs,
        DemoInk.Ink,
        style = Stroke(width = unit * 0.075f, cap = StrokeCap.Round, join = StrokeJoin.Round),
    )

    if (pose == DemoPose.Reading) {
        val origin = at(-0.20f, 0.42f)
        val book = Rect(origin, Size(unit * 0.40f, unit * 0.14f))
        drawRect(DemoInk.Paper, topLeft = book.topLeft, size = book.size)
        drawRect(DemoInk.Ink, topLeft = book.topLeft, size = book.size, style = Stroke(unit * 0.025f))
        drawLine(
            DemoInk.Ink,
            start = Offset(book.center.x, book.top),
            end = Offset(book.center.x, book.bottom),
            strokeWidth = unit * 0.02f,
        )
    }
}

File: ios/Demo/MekuriDemo/DemoFigure.swift
import SwiftUI

enum DemoPose {
    case standing
    case pointing
    case running
    case reading
}

/// A silhouette drawn from a head, a torso and stroked limbs, scaled to the
/// height it is given.
struct DemoFigure: View {
    let pose: DemoPose

    var body: some View {
        Canvas { context, size in
            let unit = size.height
            func point(_ x: CGFloat, _ y: CGFloat) -> CGPoint {
                CGPoint(x: size.width / 2 + x * unit, y: y * unit)
            }
            let limb = StrokeStyle(lineWidth: unit * 0.075, lineCap: .round, lineJoin: .round)
            let ink = GraphicsContext.Shading.color(DemoInk.ink)

            let lean: CGFloat = self.pose == .running ? 0.06 : 0
            let headRadius = unit * 0.10
            let head = CGRect(x: point(lean, 0.13).x - headRadius, y: point(0, 0.13).y - headRadius, width: headRadius * 2, height: headRadius * 2)
            context.fill(Path(ellipseIn: head), with: ink)

            var torso = Path()
            torso.move(to: point(-0.10 + lean, 0.26))
            torso.addLine(to: point(0.10 + lean, 0.26))
            torso.addLine(to: point(0.08, 0.58))
            torso.addLine(to: point(-0.08, 0.58))
            torso.closeSubpath()
            context.fill(torso, with: ink)

            var limbs = Path()
            switch self.pose {
            case .standing:
                limbs.move(to: point(-0.09, 0.30)); limbs.addLine(to: point(-0.16, 0.55))
                limbs.move(to: point(0.09, 0.30)); limbs.addLine(to: point(0.16, 0.55))
                limbs.move(to: point(-0.06, 0.58)); limbs.addLine(to: point(-0.08, 0.95))
                limbs.move(to: point(0.06, 0.58)); limbs.addLine(to: point(0.08, 0.95))
            case .pointing:
                limbs.move(to: point(-0.09, 0.30)); limbs.addLine(to: point(-0.15, 0.55))
                limbs.move(to: point(0.09, 0.30)); limbs.addLine(to: point(0.36, 0.30))
                limbs.move(to: point(-0.06, 0.58)); limbs.addLine(to: point(-0.08, 0.95))
                limbs.move(to: point(0.06, 0.58)); limbs.addLine(to: point(0.10, 0.95))
            case .running:
                limbs.move(to: point(-0.04, 0.30)); limbs.addLine(to: point(-0.24, 0.42))
                limbs.move(to: point(0.14, 0.30)); limbs.addLine(to: point(0.30, 0.18))
                limbs.move(to: point(-0.06, 0.58)); limbs.addLine(to: point(-0.30, 0.82))
                limbs.move(to: point(0.06, 0.58)); limbs.addLine(to: point(0.26, 0.92))
            case .reading:
                limbs.move(to: point(-0.09, 0.30)); limbs.addLine(to: point(-0.12, 0.46))
                limbs.move(to: point(0.09, 0.30)); limbs.addLine(to: point(0.12, 0.46))
                limbs.move(to: point(-0.06, 0.58)); limbs.addLine(to: point(-0.08, 0.95))
                limbs.move(to: point(0.06, 0.58)); limbs.addLine(to: point(0.08, 0.95))
            }
            context.stroke(limbs, with: ink, style: limb)

            if self.pose == .reading {
                let book = CGRect(x: point(-0.20, 0.42).x, y: point(0, 0.42).y, width: unit * 0.40, height: unit * 0.14)
                context.fill(Path(book), with: .color(DemoInk.paper))
                context.stroke(Path(book), with: ink, lineWidth: unit * 0.025)
                var spine = Path()
                spine.move(to: CGPoint(x: book.midX, y: book.minY))
                spine.addLine(to: CGPoint(x: book.midX, y: book.maxY))
                context.stroke(spine, with: ink, lineWidth: unit * 0.02)
            }
        }
    }
}
